docs/development_guidelines.md

```diff
@@ -76,3 +76,88 @@ but also helps fellow contributors making adjustments to your code in the future
 By default, Home Assistant will statically check for type hints in our automated CI process.
 Python modules can be included for strict checking, if they are fully typed, by adding an entry
 to the `.strict-typing` file in the root of the Home Assistant Core project.
+
+### State, argument, and return value validation
```

Comment:
I think we should remove this section. It's not specific to Home Assistant.

Reply:
Then can you provide a link to a good guide about this? Code reviewers enforce this, but first-time contributors don't know what's expected.

Reply:
What specifically do we enforce? It just looks like general Python practices to me.

Reply:
> What specifically do we enforce?

Validation of various variables where it makes sense, like checking and appropriately handling the return value of a library function call.

> It just looks like general Python practices to me.

Indeed, and you're a great developer. But not everyone is, there's many first-time or hobbyist coders trying to contribute, and code reviews would go a lot smoother if we could get these things out of the way first.

It's kind of like how CI checks for style and typing and linting so code reviewers can concentrate on more important things. By documenting what you expect of code authors, you don't have to explain it to every new person each time you review their code.

Reply:
We don't need to explain the Python language or good practices of it. That is better done elsewhere.

We have code review to make sure we achieve high quality code besides following the project specific standards. Only the project specific standards and practices should be documented.

Reply:
> We don't need to explain the Python language or good practices of it. That is better done elsewhere.

Sure, there are better resources for best practices. Let's at least link to them so people know what to follow. E.g.: here's a good guide on assertions: <https://realpython.com/python-assert-statement>, it says that assertions are good for documenting and debugging code, but bad for validating data. Here's a good style guide: <https://docs.python-guide.org/writing/style/>.

But then there's conflicting best-practice idioms too. Do we prefer [Easier to Ask Forgiveness than Permission](https://docs.python.org/3/glossary.html#term-EAFP) or [Look Before You Leap](https://docs.python.org/3/glossary.html#term-LBYL)? Do we prefer [one return point or return early](https://docs.python-guide.org/writing/style/#returning-values) from functions?

> We have code review to make sure we achieve high quality code besides following the project specific standards. Only the project specific standards and practices should be documented.

I would love some documentation about project standards for assertions and error handling but there currently isn't any.

Reply:
We'll add a sentence about the `assert` in this PR. We don't have a single standard for error handling. In service handlers we prefer raising an exception on error since the exception argument will be shown in the UI to the user.

I don't think we should add any links about general Python topics, until we see a big need for our project workflow. Links can easily become outdated. If contributors want to learn more about Python they can search for that topic online.

Reply:
Ok, I'm somewhat happy that there's going to be something written down to refer to. But, after making all of the requested edits, there isn't really anything I've contributed here any more. I'm going to close this PR unmerged so @MartinHjelmare can open one and be credited with the changes.